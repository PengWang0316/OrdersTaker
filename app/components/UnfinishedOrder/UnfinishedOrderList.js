import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  List, ListItem, ListItemText, ListItemSecondaryAction, Avatar, Typography
} from '@material-ui/core';
import { CheckCircle as CheckCircleIcon } from '@material-ui/icons';
import { withStyles } from '@material-ui/core/styles';
import { lightGreen } from '@material-ui/core/colors';

import { ITEM_ONE_PRICE_KEY } from '../../config';

const styles = {
  root: {
    width: '100%'
  },
  avatar: {
    marginRight: 10
  },
  checkIcon: {
    color: lightGreen[500]
  }
};

const getQtyString = qtyObject => {
  let returnStr = '';
  Object.keys(qtyObject).forEach(key => {
    if (key !== ITEM_ONE_PRICE_KEY) returnStr += `${key} × ${qtyObject[key]} `;
    else returnStr += `× ${qtyObject[key]}`;
  });
  return returnStr;
};

export const UnfinishedOrderList = ({
  order, menuItems, classes, onClick
}) => (
  <List className={classes.root}>
    {menuItems && Object.keys(order.items).map(key => (
      <ListItem key={key} button name={key} onClick={() => onClick(order._id, key)}>
        <Avatar src={menuItems[key].photo} className={classes.avatar} />
        <Typography color="primary">
          {menuItems[key].name}
        </Typography>
        <ListItemText primary={getQtyString(order.items[key].qty)} />
        {order.finishedItems && order.finishedItems[key] && (
          <ListItemSecondaryAction>
            <CheckCircleIcon className={classes.checkIcon} />
          </ListItemSecondaryAction>
        )}
      </ListItem>
    ))}
  </List>
);
UnfinishedOrderList.propTypes = {
  classes: PropTypes.object.isRequired,
  order: PropTypes.object.isRequired,
  menuItems: PropTypes.object,
  onClick: PropTypes.func.isRequired
};
UnfinishedOrderList.defaultProps = {
  menuItems: null
};
/* istanbul ignore next */
const mapStateToProps = state => ({
  menuItems: state.menuItems
});
export default connect(mapStateToProps, null)(withStyles(styles)(UnfinishedOrderList));
