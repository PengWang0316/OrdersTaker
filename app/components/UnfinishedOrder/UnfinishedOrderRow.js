import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Typography, CircularProgress, IconButton } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { DoneAll as DoneAllIcon, Delete as DeleteIcon } from '@material-ui/icons';
import { lightGreen, red } from '@material-ui/core/colors';

/* istanbul ignore next */
const styles = theme => ({
  root: {
    width: '100%',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  flexDiv: {
    display: 'flex',
    alignItems: 'center'
  },
  finishPercentage: {
    fontWeight: 'bold',
    size: 14
  },
  progress: {
    position: 'relative',
    left: -33
  },
  circleIcon: {
    width: 30,
    height: 30,
    color: lightGreen[700]
  },
  deleteIcon: {
    width: 20,
    height: 20
  },
  deleteBtn: {
    marginLeft: 10,
    // color: theme.palette.getContrastText(red[700]),
    // backgroundColor: red[700],
    '&:hover': {
      color: theme.palette.getContrastText(red[700]),
      backgroundColor: red[600]
    }
  },
  finishedDate: {
    marginLeft: -83
  }
});

const getPercentageOfFinishing = (finishedItemsLength, itemsLength) => Math.floor((finishedItemsLength / itemsLength) * 100);

export const UnfinishedOrderRow = ({ order, classes }) => {
  const orderDate = new Date(order.dateStamp);
  let finishingPercentage = 0;
  if (order.finishedItems) { // If the order has some finished items, calculate the percentage for the finishing.
    const finishedItemsLength = Object.keys(order.finishedItems).length;
    const itemsLength = Object.keys(order.items).length;
    finishingPercentage = finishedItemsLength === itemsLength
      ? 100 : getPercentageOfFinishing(finishedItemsLength, itemsLength);
  }
  return (
    <div className={classes.root}>
      <div className={classes.flexDiv}>
        <Typography color="primary">Table {order.tableNumber}</Typography>
        {finishingPercentage === 100 && (
          <IconButton className={classes.deleteBtn}>
            <DeleteIcon className={classes.deleteIcon} />
          </IconButton>)}
      </div>
      <Typography color="textPrimary" className={finishingPercentage === 100 ? classes.finishedDate : null}>
        {`${orderDate.getHours()}:${orderDate.getMinutes()} ${orderDate.toDateString()}`}
      </Typography>
      <div className={classes.flexDiv}>
        {finishingPercentage === 100 && <DoneAllIcon className={`${classes.circleIcon} ${classes.progress}`} />}
        {finishingPercentage !== 100 && (
          <Fragment>
            <Typography color="primary" className={classes.finishPercentage}>{finishingPercentage}%</Typography>
            <CircularProgress variant="static" value={finishingPercentage} size={36} className={classes.progress} />
          </Fragment>
        )}
      </div>
    </div>
  );
};
UnfinishedOrderRow.propTypes = {
  classes: PropTypes.object.isRequired,
  order: PropTypes.object.isRequired
};
export default withStyles(styles)(UnfinishedOrderRow);
