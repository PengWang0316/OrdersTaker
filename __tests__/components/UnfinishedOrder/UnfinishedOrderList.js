import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { UnfinishedOrderList } from '../../../app/components/UnfinishedOrder/UnfinishedOrderList';

jest.mock('@material-ui/core/List', () => 'List');
jest.mock('@material-ui/core/ListItem', () => 'ListItem');
jest.mock('@material-ui/core/ListItemText', () => 'ListItemText');
jest.mock('@material-ui/core/ListItemSecondaryAction', () => 'ListItemSecondaryAction');
jest.mock('@material-ui/core/Avatar', () => 'Avatar');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/icons/CheckCircle', () => 'CheckCircle');

describe('UnfinishedOrderList', () => {
  const defaultProps = {
    classes: {
      root: 'root',
      avatar: 'avatar',
      checkIcon: 'checkIcon'
    },
    order: {
      _id: '1',
      items: {
        123: {
          qty: {
            small: 1
          }
        }
      }
    },
    menuItems: {
      123: {
        name: '123name'
      }
    },
    onClick: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<UnfinishedOrderList {...defaultProps} />);

  test('Snapshot without the CheckCircleIcon', () => expect(renderer.create(<UnfinishedOrderList {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot with the CheckCircleIcon', () => expect(renderer.create(<UnfinishedOrderList {...{ ...defaultProps, order: { _id: '1', items: { 123: { qty: { _onePrice: 2 } } }, finishedItems: { 123: true } } }} />).toJSON()).toMatchSnapshot());

  test('Click ListItem', () => {
    const component = getShallowComponent();
    component.find('ListItem').simulate('click');
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
    expect(defaultProps.onClick).toHaveBeenLastCalledWith('1', '123');
  });
});
