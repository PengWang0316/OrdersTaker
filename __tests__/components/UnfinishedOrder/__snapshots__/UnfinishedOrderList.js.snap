// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnfinishedOrderList Snapshot with the CheckCircleIcon 1`] = `
<List
  className="root"
>
  <ListItem
    button={true}
    name="123"
    onClick={[Function]}
  >
    <Avatar
      className="avatar"
    />
    <Typography
      color="primary"
    >
      123name
    </Typography>
    <ListItemText
      primary="× 2"
    />
    <ListItemSecondaryAction>
      <CheckCircle
        className="checkIcon"
      />
    </ListItemSecondaryAction>
  </ListItem>
</List>
`;

exports[`UnfinishedOrderList Snapshot without the CheckCircleIcon 1`] = `
<List
  className="root"
>
  <ListItem
    button={true}
    name="123"
    onClick={[Function]}
  >
    <Avatar
      className="avatar"
    />
    <Typography
      color="primary"
    >
      123name
    </Typography>
    <ListItemText
      primary="small × 1 "
    />
  </ListItem>
</List>
`;
